app/services/notify_service.rb

```diff
@@ -196,6 +200,24 @@ def call(recipient, type, activity)
 
   private
 
+  def notification_group_key
```

Comment:
A good chunk of this method is dealing with attributes on `Notification`, and the parts that are not are managing the hour bucket concept and how it interacts with redis. I wonder if some/all of:

- Move this method onto `Notification` directly? (may need to pass in the recipient, as currently done)
- Add a class which manages the redis querying/checking/setting/key aspects?
- Keep this in the notify service, but turn this one larger method into 2-3 smaller methods which separate the different things being done here